text/0000-else.md

```diff
@@ -0,0 +1,269 @@
+
+In the case of LWC, since we already have `if:else` and `if:false`, and since there is no ergonomic way to do conditional rendering in JavaScript (other than `render()`, which requires creating separate template HTML files), it would be most natural to align with the first group of frameworks.
+
+## Detailed design
+
+The basic design is to add three new directives:
+
+- `lwc:if`
+- `lwc:elseif`
+- `lwc:else`
+
+The reason for not reusing the existing `if:true` or `if:else` is:
+
+1. It is more consistent to use `lwc:*` for all three.
+2. There is no need for `*:false`, assuming we allow `!` in template expressions, e.g. `lwc:if={!foo}`.
+
+### Syntax
+
+Both `lwc:elseif` and `lwc:else` must be immediately preceded by a sibling `lwc:if` or `lwc:elseif`.
+
+```html
+<!-- Valid -->
+<template lwc:if={foo}></template>
+<template lwc:else></template>
+```
+
+```html
+<!-- Invalid -->
+<template lwc:if={foo}></template>
+<div>Yolo!</div>
+<template lwc:else></template>
+```
+
+Whitespace is ignored when considering siblinghood.
+
+Furthermore, these directives (along with `if:true` and `if:false`) are mutually exclusive, and multiple cannot be applied to the same element:
+
+```html
+<!-- Invalid -->
+<template lwc:if={foo} if:true={foo}></template>
+```
+
+```html
+<!-- Invalid -->
+<template lwc:if={foo}></template>
+<template lwc:elseif={bar} lwc:else></template>
+```
+
+The existing `if:true` and `if:else` directives _cannot_ be combined with `lwc:elseif` or `lwc:else`:
+
+```html
+<!-- Invalid -->
+<template if:true={foo}></template>
+<template lwc:elseif={bar}></template>
+```
+
+```html
+<!-- Invalid -->
+<template if:true={foo}></template>
+<template lwc:else></template>
+```
+
+#### Code comments
+
+In the case of code comments, the behavior depends on whether `lwc:preserve-comments` is enabled or not. When _not_ preserving comments, comments may appear between conditional siblings:
+
+```html
+<!-- Valid -->
+<template>
+    <template lwc:if={foo}></template>
+    <!-- Comment! -->
+    <template lwc:else></template>
+</template>
+```
+
+Whereas when `lwc:preserve-comments` is enabled, comments become syntactically meaningful and therefore cannot be placed between sibling conditional directives:
+
+```html
+<!-- Invalid -->
+<template lwc:preserve-comments>
+    <template lwc:if={foo}></template>
+    <!-- Comment! -->
+    <template lwc:else></template>
+</template>
+```
+
+#### Accessors
+
+Similar to `if:true` and `if:false`, the expression passed in to `lwc:if` and `lwc:elseif` must use [simple dot notation](https://lwc.dev/guide/reference#if%3Atrue%7Cfalse%3D%7Bexpression%7D).
+
+The exception is that "dot notation" is now expanded to include expressions preceded by a single `!` character, indicating negation:

```

Comment:
`{!` is the delimeter for expressions in Salesforce: [Aura](https://developer.salesforce.com/docs/atlas.en-us.234.0.lightning.meta/lightning/expr_overview.htm), [Visualforce](https://trailhead.salesforce.com/content/learn/modules/visualforce_fundamentals/visualforce_variables_expressions#:~:text=The%20expression%20syntax%20in%20Visualforce%20is:). 

This may cause confusion. I think some user research is needed.

Reply:
This is a good callout. On this topic, I think it is also worth adding that in all the modern JS frameworks, using the `!` token is used as the negation operator in template expressions.

It appears to me that we don't have a consensus on the template expression syntax extension proposed in this RFC. Taking a step back the introduction of the `!` operator is orthogonal to if/else proposal. I would suggest removing it from this proposal and moving the discussion to another RFC. The current lack of capability to express negation in the template makes the use of the if/else new directive less friendly since it would require developers to express the negation in JavaScript, but I think we can work on this in parallel.

Thoughts @nolanlawson @caridy?

Reply:
If we get rid of `!`, then we should revert https://github.com/salesforce/lwc-rfcs/pull/62/commits/0639cea2cb6f87db8241613e5f215f25ab129c67 and go back to the `elseif:true` / `elseif:false` / `lwc:if` system.

**Pros:** No need for the `!` in expressions. We can reuse the existing `if:true` / `if:false` rather than deprecating them.

**Cons:** The new system is less internally consistent compared to everything being prefixed with `lwc:`.

I am fine either way. Internal consistency is not everything. 🙂

Reply:
We could also rename `lwc:else` to `else:render`. Maybe this would look a bit more consistent:

```html
<template if:true={foo}></template>
<template elseif:true={bar}></template>
<template else:render></template>
```

We can also bikeshed on the name. (`else:do`? `else:show`?)

Reply:
Just chatted with @pmdartus ; we have a new proposal:

1. Get rid of the `!`
2. Keep `lwc:if` / `lwc:elseif` / `lwc:else`
3. Move the discussion about complex expressions (including `!`) to a future proposal

The benefit is that we can unblock this proposal for the time being, and consider complex expression syntax later.